### DISCUSSION TITLE
🎉 New Feature: Quick tools window (v1.12-BETA)

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#124] [Announcements]
[Install the beta via BRAT](https://github.com/chrisgurney/obsidian-note-toolbar?tab=readme-ov-file#installation) and let me know what you think!

## New Feature 🎉

**Quick tools window**

Introducing a new, convenient way to access all of your toolbar items!

- To access use the `Note Toolbar: Open quick tools` command.
- Simply select the item and it will use it as if you triggered it from the toolbar.
- Map a hotkey to the command, or add to your existing toolbars for access to the kitchen sink!
- Note that menus are not shown in the list.
- _Thank you @Glint-Eye for the idea!_

![CleanShot 2024-09-02 at 08 37 21](https://github.com/user-attachments/assets/c6d4c7cc-67c0-42ce-8ac1-dbbf75fd7bb9)

## Changes

- Infrastructure for translations is underway. ([Help wanted](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/118))

## Fixes

- Settings: Fixed case for previews where icon is not available (e.g., plugin with custom icon is disabled).
 
<hr /><em>This discussion was created from the release <a href='https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.12-beta-1'>v1.12-beta-1</a>.</em>

**Comment by chrisgurney** [comment]
@FelipeRearden I'd love to hear what you think!

**Reply by FelipeRearden** [reply]
Hello @chrisgurney !!!!

First of all, forgive me for my absence in the latest beta cycle. August was a busy month for me and I could not find energy to be with you guys in the beta cycle. I couldn’t even update to the lastest version yet.

My relationship with NTB keeps getting better in a way that I NTB now is internalized into my workflow in a way that makes very happy!

> @FelipeRearden I'd love to hear what you think!

Before anything else, I can confirm that I will be a tester for version `1.12` on iPadOS using a keyboard.

I loved the idea and I am watching the intro video for almost 5 minutes and ideas are 🤯. 

I am gonna update to the beta in my test vault and start getting insights. 

Let me just ask a question before (I think you already know what I am gonna ask 😀)
- Did you target the NTB modal in a specific class to let the user customize the NTB modal without hurting the others modals?

In other words: the same way that we can style the NTB menus will be possible to do this with the NTB Modal?

**Reply by chrisgurney** [reply]
Thanks @FelipeRearden !

My first thought was not to style it differently as I'm using a built-in Obsidian component, but you're right, as I think about it as a toolbar in itself, it should be configurable and adopt the toolbar's styles by default. 

I'll add it to my list!

**Reply by chrisgurney** [reply]
BTW no pressure on testing betas, as my goal is to work towards smaller feature releases -- actually following [semantic versioning](https://semver.org/) -- as my free time disappears.

Ideally I only want to put features in beta if necessary. For example, for testing on mobile devices, gathering feedback, and for bigger changes to infrastructure.

(The last release had a quick beta really just for a mobile test.)

**Reply by FelipeRearden** [reply]
> My first thought was not to style it differently as I'm using a built-in Obsidian component, but you're right, as I think about it as a toolbar in itself, it should be configurable and adopt the toolbar's styles by default.

I was able to open console and I saw that you already isolated 🎉

![IMG_1292](https://github.com/user-attachments/assets/1b0b988b-4197-48b5-9f53-99de99b08c06)

From my experience on customizing modals from plugins. It 100% customizable in `beta-4` I did some play and is perfect this way :) 

> BTW no pressure on testing betas, as my goal is to work towards smaller feature releases…

Okay, thanks !
I will adjust my pace.

This is the type of feature that has a lot of different directions: my mission is to share with you the directions that I would like to implement in my workflow for you to decide  if is a good fit for NTB.

**Reply by chrisgurney** [reply]
I'm open if you think about what style settings might make sense in this dialog.

Off hand, can use the same defaults perhaps:
- icon color
- text color
- hover color
- modal color = toolbar background by default? Transparency might be at play here

**Reply by FelipeRearden** [reply]
For text color and icon color, one suggestion might be use `nth-of-type(odd)`and `nth-of-type(even)` for users with a lot of items

> modal color = toolbar background by default? Transparency might be at play here

100% agree with you. I think we could let this for css snippets from power users

**Reply by chrisgurney** [reply]
@FelipeRearden Thanks for the ideas. I've added a few styles that can be changed via the Style Settings plugin (and renamed the class on the dialog), in [this release](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.12-beta-5).

**Reply by FelipeRearden** [reply]
Thank you very much @chrisgurney  for the renamed classes!!! 

Working perfect on version `beta-6` on Obsidian Mobile.  ☀️

**Comment by FelipeRearden** [comment]
> @FelipeRearden I'd love to hear what you think!

@chrisgurney 

## FB#1 beta-4: New Idea

There are two types of NTB users:
1- Normal users
2- Complex users: 
- A lot of toolbars
- A lot of items from these toolbars
- A lot of mapping to make sure that each note/folder has the perfect NTB for the job

This Idea from this feature is 100% based on the way that would like to use this feature.

**The summary for this feature is to be able to MAP the items inside Quick Actions based on Toolbars (more than one and LESS then all)**
- Complex users does not need to use in a modal ALL NTB items. They need to use items based on a criteria.

### Let me give you a real example: 

For processing content from books, I have a NTB toolbar with 5 items (3 + 2 Menus) 
- 1 Menu with 3 items 
- another is a Menu with 10 items

The Menu with 10 items are commands that I dont have to use every time that I am writing. 

They are settings commands related to the books workflow that needs to be around, I NEED TO HAVE A MENU FOR THESE COMMANDS to have them around.

That is where the Quick Actions could be a game changer.
- Instead a Menu with 10 commands, I could transform this NTB Menu item in a NTB Modal and remove than from this Book Toolbar.

<hr>

Now imagine this same rule applied to 10 different workflows with taylor made Quick Actions build for each one with only a global hotkey to trigger the modal.

This a dream for Complex users. 

<hr>

the Quick Actions Settings  could use the Folder Map UI by adding a third column to the mapping
- the only difference is that it should be a multi-selection to be able to select more than one toolbar

![IMG_1293](https://github.com/user-attachments/assets/5290f3a0-8a16-40c8-b855-45ea59ac4253)


<hr> 

let me know what you think, them I can elaborate more about this.

I am gonna keep testing looking for errors and for new ideas ☀️

**Reply by chrisgurney** [reply]
Interesting idea. 

I will be making the mappings lines UI expandable (like toolbar items), in order to support settings for [more granular toolbars](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/120#discussioncomment-10639663).

I'll keep this in mind as I hear feedback on the latest release. Thank you!

**Reply by FelipeRearden** [reply]
> Interesting idea.

I am happy to know that you found an interesting idea!

I think the main issue from power NTB user will be the amount of items that we have in our databases. Be able to filter these new `granular toolbars` will be a huge feature for users  

> I'll keep this in mind as I hear feedback on the latest release. Thank you!

☀️

**Comment by FelipeRearden** [comment]
### FB#3 beta-8/9

> Quick Tools window now supports opening File items via keyboard:
> - in a new tab via cmd or ctrl + enter
> - in a split via cmd+option or ctrl+alt + enter
> 
> Toolbar Focus mode now supports opening File items via keyboard via the same keys, but only if that hotkey is not already mapped.

Both features (Quick Tools and NTB Focus command) are working perfect on `beta-9` on Obsidian Mobile. I tested with the scenarios with and without hotkeys for `cmd+option+enter` and `cmd+enter`

<hr>

ℹ️

This is a comment about a behavior that I saw on `beta-8` and `beta-9` that **I am not 100% sure** because I was not using the `NTB Focus` command on this beta cycle.

Sometimes executing the command from the keyboard using hotkeys the NTB was not VISUAL FOCUSED (border around the focused item) but It was FOCUSED since the focus was out of the active note and  in the NTB (I know because hitting ENTER executed the NTB item command)

I am just adding this comment beacause I dont know if something changed in `beta-8/9` code related to  the Focus.

I am gonna keep checking on next betas this behavior or in the public release because this is the type of thing that is good to test in real world workflow where I hit this command many times per hour

**Reply by chrisgurney** [reply]
@FelipeRearden Hmm, is it possible to show a video? Or at least share what the configuration of the toolbar was? I just need to see the first few items, with their types + visibility, to make sure it's getting the right first item to focus on.

**Reply by FelipeRearden** [reply]
I am creating the screen record to send to you and there is no bug. Before I push REC there was a bug 🤯
- I send the video to you when I get

This is the NTB 🚨 Look at the end the Drop-down for **Mobile** this `Set position:` prefix does not exist on the pubic version 
![IMG_1298](https://github.com/user-attachments/assets/72891606-3597-4f74-b350-97582947c049)


This is the  `data.json`  

```

  "folderMappings": [
    {
      "folder": "*",
      "toolbar": "b6eceff8-49a7-4706-96e5-c52670e5c97e"
    },
    {
      "folder": "Folder A/Note A",
      "toolbar": "b6eceff8-49a7-4706-96e5-c52670e5c97e"
    },
    {
      "folder": "Folder A",
      "toolbar": "1c137d9c-87a0-453b-843d-3c757b0ad51e"
    }
  ],
  "icon": "circle-ellipsis",
  "showEditInFabMenu": false,
  "toolbarProp": "notetoolbar",
  "toolbars": [
    {
      "defaultStyles": [
        "border",
        "even",
        "sticky"
      ],
      "items": [
        {
          "label": "4",
          "icon": "lucide-album",
          "link": "Split right",
          "linkAttr": {
            "commandId": "workspace:split-vertical",
            "hasVars": false,
            "type": "command"
          },
          "tooltip": "",
          "visibility": {
            "desktop": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "mobile": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "tablet": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            }
          },
          "uuid": "c489632f-fe32-44d7-98a0-1bd97cf15131"
        },
        {
          "label": "44",
          "icon": "lucide-align-justify",
          "link": "Toggle bold",
          "linkAttr": {
            "commandId": "editor:toggle-bold",
            "hasVars": false,
            "type": "command"
          },
          "tooltip": "",
          "visibility": {
            "desktop": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "mobile": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "tablet": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            }
          },
          "uuid": "dad1d80e-0047-402b-996a-e2374fb7ea2f"
        },
        {
          "label": "444",
          "icon": "lucide-alert-triangle",
          "link": "2e3018f5-4232-4ebd-b8a5-52a3eeccf22f",
          "linkAttr": {
            "commandId": "",
            "hasVars": false,
            "type": "menu"
          },
          "tooltip": "",
          "visibility": {
            "desktop": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "mobile": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "tablet": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            }
          },
          "uuid": "3b1f9be5-326f-4055-8c97-7c3365b5dc4b"
        },
        {
          "label": "4444",
          "icon": "lucide-alarm-plus",
          "link": "1c137d9c-87a0-453b-843d-3c757b0ad51e",
          "linkAttr": {
            "commandId": "",
            "hasVars": false,
            "type": "menu"
          },
          "tooltip": "",
          "visibility": {
            "desktop": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "mobile": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "tablet": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            }
          },
          "uuid": "99c43bd9-3943-4fb3-9588-fb6650ea5c79"
        },
        {
          "uuid": "5619f5d4-1147-4e44-ac99-6648920d49d9",
          "label": "44444",
          "icon": "lucide-hash",
          "link": "other note.md",
          "linkAttr": {
            "commandId": "",
            "hasVars": false,
            "type": "file"
          },
          "tooltip": "",
          "visibility": {
            "desktop": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "mobile": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            },
            "tablet": {
              "allViews": {
                "components": [
                  "icon",
                  "label"
                ]
              }
            }
          }
        }
      ],
      "mobileStyles": [],
      "name": "l1",
      "position": {
        "desktop": {
          "allViews": {
            "position": "top"
          }
        },
        "mobile": {
          "allViews": {
            "position": "top"
          }
        },
        "tablet": {
          "allViews": {
            "position": "top"
          }
        }
      },
      "updated": "2024-09-07T13:45:09.871Z",
      "uuid": "b6eceff8-49a7-4706-96e5-c52670e5c97e"
    },
```

Let me know if you need more info. I think is something on my end due to a messy vault

**Reply by chrisgurney** [reply]
Thanks, will try to reproduce when I get a chance with similar items.

**Reply by chrisgurney** [reply]
So I have been able to reproduce on my iPad, but not on my macOS laptop using mobile emulation. I'll keep investigating.

Thank you for noting the bug with the mobile position dropdown → I was using the wrong string. Will be fixed in the next beta.

**Reply by chrisgurney** [reply]
@FelipeRearden So I think I've fixed it in [1.12-beta-10](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.12-beta-10)... but it reintroduces a bug where the focus effect sticks after using an item in the toolbar, at least on mobile. (When I fixed _that_ issue is when it stopped showing the focus effect, I believe. [Here's the commit](https://github.com/chrisgurney/obsidian-note-toolbar/commit/662b349488d0fc42dc748028fd1b22f6e2bc8d77) on July 17.)

A couple thoughts:
- Doing the `editor:focus` as you had suggested would move focus back to the editor before executing the command.
  - But I'm wondering if there's cases where users will want to execute a command multiple times by hitting enter over and over. Makes me think that changing the focus back should be an option, on an item-by-item basis.
  - If you look at the Editing Toolbar plugin, they don't execute `editor:focus` for all commands ([example](https://github.com/PKM-er/obsidian-editing-toolbar/blob/2115f3735b5bf7f5586cb72164481a7fd54c54fa/src/plugin/main.ts#L379), which you can compare to the command above it).
- Setting a class on the currently focussed item when using the NTB Focus command should also work. I'm thinking of taking this approach for now, to fix this particular bug.

**Reply by chrisgurney** [reply]
@FelipeRearden Working on the solution mentioned. Hopefully will have another beta for you soon to test, once I work out some edge cases (with the focus effect getting stuck).

**Reply by FelipeRearden** [reply]
Hello @chrisgurney !!!!

> Thank you for noting the bug with the mobile position dropdown → I was using the wrong string. Will be fixed in the next beta.
> fix: item focus not using user-defined border radius
> fix: using wrong strings in position dropdown 

Working perfect on `beta-10` on Obsidian Mobile on iPad. Thank you for the quick fix 🙏

> Doing the editor:focus as you had suggested would move focus back to the editor before executing the command.
> 
>  But I'm wondering if there's cases where users will want to execute a command multiple times by hitting enter over and over. Makes me think that changing the focus back should be an option, on an item-by-item basis.
> 
>  If you look at the Editing Toolbar plugin, they don't execute editor:focus for all commands ([example](https://github.com/PKM-er/obsidian-editing-toolbar/blob/2115f3735b5bf7f5586cb72164481a7fd54c54fa/src/plugin/main.ts#L379), which you can compare to the command above it).
> 
> Setting a class on the currently focussed item when using the NTB Focus command should also work. I'm thinking of taking this approach for now, to fix this particular bug.

I never thought about the repetitive workflow before, and I 100% agree that we have this user case (keep focus on NTB item after execution). It is not my workflow, but it is feasible. 

Based on this, I have one idea to share with you. 

1️⃣ On `Command` type you could use the same logic for `ENTER` and `CMD+ENTER` that you used on `File` type to get both scenarios:

- `ENTER` will get the focus on active note
- `CMD + ENTER` will get the focus back on the NTB item

And a global setting on Settings to change the behavior between `ENTER` and `CMD + ENTER`

Just an idea, maybe this approach could save you same time and satisfy both worlds at the same time 🤔

<hr>

> Working on the solution mentioned. Hopefully will have another beta for you soon to test, once I work out some edge cases (with the focus effect getting stuck).

Okay, I will keep my eyes for the next beta test cycle 

I wish you a fantastic day and week ☀️

**Reply by chrisgurney** [reply]
@FelipeRearden Can you give [1.12-beta-11](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.12-beta-11) a go when you get a chance with the `Focus` command?

I've also changed how search results are displayed; not sure if that will impact how you use the dialog -- though it sounds like you're going to completely hide it?

**Reply by FelipeRearden** [reply]
> @FelipeRearden Can you give [1.12-beta-11](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.12-beta-11) a go when you get a chance with the Focus command?

Working perfect on `beta-12` 🎉 

I am gonna keep testing `focus command` for the next releases and after the public release. 

Look perfect now and I tested again the new keyboard features from `beta-8/9` and they are working perfect as well. 

> I've also changed how search results are displayed; not sure if that will impact how you use the dialog -- though it sounds like you're going to completely hide it?

It is a great idea and it will help users to get the exact item that they are looking for! 

The workflow that described in the video on #126 100% depends on the NOT FUZZY feature because I really need to be able to return the exact searchable term (single term) to make this workflow fully useful. 
- I am  in a 11 inches iPad screen and I need to take care of every space that I can 😀

If it was not feasible, I will rethink my approach for Quick Tools because I am in love with this modal and I really wanna have a workflow around this amazing UI 🙏

**Reply by chrisgurney** [reply]
What if there was a similar dialog that let you search for toolbar first, then once selected, you can search through that toolbar's items?

I'm not 100% clear on your desired workflow, but would that help? Just thinking through options that might benefit the most users.

**Reply by FelipeRearden** [reply]
> What if there was a similar dialog that let you search for toolbar first, then once selected, you can search through that toolbar's items?
> 
> I'm not 100% clear on your desired workflow, but would that help? Just thinking through options that might benefit the most users.

After some hours struggling, I was able to find a solution using CSS. I explain in the last post on #126 . 🎉

Thank you very much for the help and the attention. I am happy that I could achieve this without compromising the core idea of Quick Tools 🙏

**Comment by FelipeRearden** [comment]
Hello @chrisgurney !!!

## FB#4 beta-16

> Open with the Note Toolbar: Open quick tools (toolbar) command.
> 
> Pressing ← in the Quick Tools window when the item search is empty will navigate back to the Quick Toolbars window.

Both features working perfect on `beta-16` on Obsidian Mobile on iPadOS.

> Quick Tools window now supports opening File items via keyboard:
> 
> in a new tab via cmd or ctrl + enter
> in a split via cmd+option or ctrl+alt + enter

Still working perfect when executing from the new modal (quick tools (toolbar)) as well

☀️ `Focus` command working perfect on `beta-16` as well. 🙏

**Comment by FelipeRearden** [comment]
## FB#4 beta-16 | New Idea ONLY for Quick Tools Toolbar

The idea is simple: since we have now a new modal for the Toolbars, I think we need to be able to add an ICON to a toolbar the same way that we do for NTB items.

ICONS are the signature of the NTB plugin. It is strange to open the Modal and not see an icon 🥲

<hr>

I am gonna wait for your feedback on this idea. Because … I have a BIG idea related to be abe to add ICONS to a Toolbar which could be extreme useful and could add a new meaning to the toolbars without much development effort.

**Reply by chrisgurney** [reply]
@FelipeRearden Hmm, I haven't seen a need to have icons for toolbars yet, but if there's an idea for how they might be used (aside from making the toolbar selector look pretty), I'm interested in hearing about it!

BTW I did try adding small toolbar previews under each toolbar name in that window, but it seemed to noisy without adding any real value.

**Reply by FelipeRearden** [reply]
> if there's an idea for how they might be used (aside from making the toolbar selector look pretty), I'm interested in hearing about it.

### Main Idea

The idea is to have an icon that represents the NTB from the Active Note in the statusbar.
- If I change to another active note that has a different NTB, the icon in the toolbar will change as well.

📽️ Watch in the bottom right how the first icon in the status **changes when I change the active note.**

https://github.com/user-attachments/assets/e5baf7d6-bb31-4ab5-8046-172e18aa1b81


This is the behavior that I am thinking here for a toolbar icon in the status bar.

### Principles

This ICON has two objectives: 

🅰️ Give to the user a visual signal of something about the active note that is important for the workflow.

🅱️ When click in the icon, it will execute a command defined by the user in the toolbar settings. 

### Details

1️⃣ No keyboard navigation

2️⃣ the command executed by the icon the statusbar will be different for each toolbar. In this scenario, any toolbar will gain a new usage.

3️⃣ Disconnection/independence between the commands from the items in the toolbar and the command from the toolbar

<hr>

the idea is to use the NTB structure to create a dynamic way to execute some command in the vault. Plus, using the statusbar combined with visual representation which is presented on both NTB and statusbar will give the user one more tool to automate the active note.

The status bar is a place almost never touched by plugins and is under used in my opinion. The idea here is to using the NTB ecosystem to bring life to a place that is always in our screens. 


Let me know what you think ☀️

**Reply by chrisgurney** [reply]
@FelipeRearden I'm not convinced about putting it in the status bar, as that's not really the intent of that part of the UI. That said, I've seen various controls put down there by other plugins, so I'm not _totally_ against the idea.

I _have_ thought some more about icons for toolbars, and did come up with a few places it would help:

- floating button
- as the default for Menu items (if the user hasn't set an icon yet)
- file menu (coming to the next beta)

Thinking out loud here: My thought is to add it to the toolbar settings towards the _bottom_ of the window, perhaps, as it's completely optional, and I don't want it to get in the way of users quickly creating new toolbars.

**Reply by FelipeRearden** [reply]
> I've seen various controls put down there by other plugins, so I'm not totally against the idea.

🙏 🙏 🙏  As I said in my comment, it is the type of feature for users who are looking for elegant and organic ways to add more automation to their workflows without hurting the screen with new modals, containers and hotkeys. 

A dynamic command with a dynamic UI: The icon changes based on the ntb config and clinking on it will execute a command.  

> I have thought some more about icons for toolbars, and did come up with a few places it would help:

3 amazing places for the icons. Specially File Menus and Menu items.

> Thinking out loud here: My thought is to add it to the toolbar settings towards the bottom of the window, perhaps, as it's completely optional, and I don't want it to get in the way of users quickly creating new toolbars.

10000% agree with you. This way the user decides about this or not and gives us more freedom to implement features around the toolbar without worry about hurting Toolbar items UI.

**Comment by FelipeRearden** [comment]
@chrisgurney 

As always, I am gonna do my test on the plubic release of version `1.12` .

Before we finish the info about `1.12`, let me know if you want me to create `new issues` about the 2 ideas from this beta cycle:

1- Folder Map for Quick Tools
2- Active Toolbar inside the status bar

**Reply by chrisgurney** [reply]
Sure, please do (create the issues). Might be useful to link from other similar discussions and further flesh them out. Thanks!

**Reply by FelipeRearden** [reply]
> 1- Folder Map for Quick Tools

#141
